File: apps/android/app/src/main/java/com/exponential/app/ui/home/BoardSwitcherSheet.kt
package com.exponential.app.ui.home

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.key
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.exponential.app.data.db.ServerBoardGroup
import com.exponential.app.data.db.TeamBlock
import com.exponential.app.domain.CodingSessionDisplayState
import com.exponential.app.domain.PinnedRow
import com.exponential.app.domain.codingSessionDisplayState
import com.exponential.app.ui.components.BoardRow
import com.exponential.app.ui.components.TeamAvatar
import com.exponential.app.ui.components.GlassSheet
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.issue.DoneBlue
import com.exponential.app.ui.issue.LiveDot
import com.exponential.app.ui.issue.NeedsInputAmber
import com.exponential.app.ui.issue.ReviewGreen
import com.exponential.app.ui.issue.StaticDot
import com.exponential.app.ui.session.SessionRowTitle
import com.exponential.app.ui.session.sessionRowIdentifier
import com.exponential.app.ui.session.sessionRowTitle
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassRow

/**
 * The inline board switcher: a bottom sheet presenting every signed-in
 * account's teams and boards (server → team → board). This is
 * the old Boards home screen's tree, relocated — picking a board swaps the
 * Issues tab's list in place instead of pushing a new destination.
 * Mirrors iOS BoardSwitcherSheet. EXP-698 r5 gave every team block its own
 * "Create board" row (the target team is the block's, so nothing is ambiguous
 * any more) and put a "New team" row at the very bottom.
 */
@Composable
fun BoardSwitcherSheet(
    groups: List<ServerBoardGroup>,
    /** The board the Issues tab is showing — its row takes the active paint. */
    currentBoardId: String?,
    onSelect: (accountId: String, boardId: String) -> Unit,
    onCreateBoard: (teamId: String) -> Unit,
    onCreateTeam: () -> Unit,
    onDismiss: () -> Unit,
    /** EXP-778: the active team's resolved pins, in sort_order; empty hides
     *  the section. The sheet is the mobile sidebar, so "Pinned" sits at its
     *  top like the web/desktop rail group. */
    pinned: List<PinnedRow> = emptyList(),
    onOpenPin: (PinnedRow) -> Unit = {},
) {
    GlassSheet(title = "Switch board", onDismiss = onDismiss) {
        if (groups.isEmpty() && pinned.isEmpty()) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 40.dp, vertical = 32.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                Icon(
                    ExpIcons.navBoards,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                    modifier = Modifier.size(22.dp),
                )
                Text(
                    "Create your first board on the web or desktop app.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                    textAlign = TextAlign.Center,
                )
            }
        } else {
            LazyColumn(
                contentPadding = PaddingValues(start = 16.dp, end = 16.dp, top = 4.dp, bottom = 24.dp),
                verticalArrangement = Arrangement.spacedBy(18.dp),
                modifier = Modifier.fillMaxWidth(),
            ) {
                if (pinned.isNotEmpty()) {
                    item(key = "pinned") {
                        PinnedSection(rows = pinned, onOpen = onOpenPin)
                    }
                }
                items(groups, key = { it.accountId }) { group ->
                    ServerSection(
                        group = group,
                        showServerHeader = groups.size > 1,
                        currentBoardId = currentBoardId,
                        onSelect = onSelect,
                        onCreateBoard = onCreateBoard,
                    )
                }
                // The last row in the sheet, under every server's teams: a
                // team is the one thing the tree above can't offer to make.
                item(key = "new-team") {
                    MutedActionRow(
                        label = "New team",
                        testTag = "board-switcher-new-team",
                        onClick = onCreateTeam,
                    )
                }
            }
        }
    }
}

@Composable
private fun ServerSection(
    group: ServerBoardGroup,
    showServerHeader: Boolean,
    currentBoardId: String?,
    onSelect: (accountId: String, boardId: String) -> Unit,
    onCreateBoard: (teamId: String) -> Unit,
) {
    Column(modifier = Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(8.dp)) {
        // The server/email header only disambiguates when several accounts are
        // signed in — with a single account it's noise (iOS parity).
        if (showServerHeader) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(horizontal = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        group.hostname,
                        style = MaterialTheme.typography.titleSmall,
                        color = MaterialTheme.colorScheme.onSurface,
                    )
                    if (!group.userEmail.isNullOrBlank()) {
                        Text(
                            group.userEmail,
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                        )
                    }
                }
            }
        }
        group.teamBlocks.forEach { block ->
            TeamBlockView(
                accountId = group.accountId,
                block = block,
                currentBoardId = currentBoardId,
                onSelect = onSelect,
                onCreateBoard = onCreateBoard,
            )
        }
    }
}

@Composable
private fun TeamBlockView(
    accountId: String,
    block: TeamBlock,
    currentBoardId: String?,
    onSelect: (accountId: String, boardId: String) -> Unit,
    onCreateBoard: (teamId: String) -> Unit,
) {
    Column(modifier = Modifier.fillMaxWidth(), verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            TeamAvatar(block.team, size = 18.dp)
            Spacer(Modifier.width(8.dp))
            // EXP-698 r5: no board COUNT — the rows below are the count,
            // and the number read as an unread badge.
            Text(
                block.team.name,
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurface,
                modifier = Modifier.weight(1f),
            )
        }
        block.boards.forEach { board ->
            key(board.id) {
                BoardRow(
                    board = board,
                    active = board.id == currentBoardId,
                    onClick = { onSelect(accountId, board.id) },
                )
            }
        }
        MutedActionRow(
            label = "Create board",
            testTag = "board-switcher-create-board",
            onClick = { onCreateBoard(block.team.id) },
        )
    }
}

/**
 * The switcher's two make-something rows. Deliberately NOT carded: a board row
 * is a thing you can switch to, and drawing "Create board" the same way would
 * have made the list of boards look one longer than it is.
 */
@Composable
private fun MutedActionRow(label: String, testTag: String, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            // The capture suites address these two by tag — their labels also
            // read as ordinary content elsewhere on the sheet.
            .testTag(testTag)
            .clickable(onClick = onClick)
            .padding(horizontal = GlassTokens.RowPaddingH, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            ExpIcons.uiAdd,
            contentDescription = null,
            modifier = Modifier.size(16.dp),
            tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
        )
        Spacer(Modifier.width(10.dp))
        Text(
            label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
        )
    }
}

/**
 * EXP-778: the "Pinned" group — the caller's own pins in the active team, each
 * row drawn like the list it came from: an issue as identifier + title, a
 * coding session with the Agent page's state dot + identity line, an action
 * as its glyph + name. Tapping opens the target (issue detail, the steering
 * screen, the composer with the action preselected).
 */
@Composable
private fun PinnedSection(rows: List<PinnedRow>, onOpen: (PinnedRow) -> Unit) {
    Column(
        modifier = Modifier.fillMaxWidth().testTag("board-switcher-pinned"),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                ExpIcons.uiPin,
                contentDescription = null,
                modifier = Modifier.size(14.dp),
                tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            )
            Spacer(Modifier.width(8.dp))
            Text(
                "Pinned",
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
        rows.forEach { row ->
            key(row.pin.id) {
                PinnedRowView(row = row, onClick = { onOpen(row) })
            }
        }
    }
}

@Composable
private fun PinnedRowView(row: PinnedRow, onClick: () -> Unit) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .glassRow()
            .clickable(onClick = onClick)
            .padding(horizontal = GlassTokens.RowPaddingH, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        when (row) {
            is PinnedRow.Issue -> {
                Text(
                    row.issue.identifier,
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                    fontFamily = FontFamily.Monospace,
                )
                Spacer(Modifier.width(10.dp))
                Text(
                    row.issue.title.ifBlank { "Untitled issue" },
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
            }
            is PinnedRow.Session -> {
                // The Agent page's row look (AgentSessionsList): the shared
                // identity line with the state dot in front. No device join
                // here — the sheet is a launcher, the list carries the rest.
                val state = codingSessionDisplayState(row.session, row.issue?.prState ?: row.session.prState)
                SessionRowTitle(
                    identifier = sessionRowIdentifier(row.issue),
                    title = sessionRowTitle(row.session, row.issue),
                    modifier = Modifier.weight(1f),
                    dot = {
                        when (state) {
                            // EXP-848: pulses only mid-turn (agent_busy).
                            CodingSessionDisplayState.Running ->
                                LiveDot(busy = row.session.agentBusy)
                            CodingSessionDisplayState.NeedsInput -> StaticDot(NeedsInputAmber)
                            CodingSessionDisplayState.Review -> StaticDot(ReviewGreen)
                            CodingSessionDisplayState.Done -> StaticDot(DoneBlue)
                        }
                    },
                )
            }
            is PinnedRow.Action -> {
                Icon(
                    row.action.icon?.takeIf { it.isNotEmpty() }?.let { ExpIcons.byName(it) }
                        ?: ExpIcons.actionDefault,
                    contentDescription = null,
                    modifier = Modifier.size(16.dp),
                    tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
                )
                Spacer(Modifier.width(10.dp))
                Text(
                    row.action.name,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.weight(1f),
                )
            }
        }
        Spacer(Modifier.width(8.dp))
        Icon(
            ExpIcons.uiChevronRight,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Quaternary),
            modifier = Modifier.size(16.dp),
        )
    }
}

File: apps/ios/Exponential/UI/Home/BoardSwitcherSheet.swift
import ExpUI
import ExpCore
import SwiftUI

/// Bottom-sheet board picker for the Issues tab's inline switcher: the
/// server → team → board tree that used to be the Boards overview
/// screen, now presented modally. Selecting a board swaps the Issues tab's
/// list in place (the caller writes last-used and dismisses).
/// EXP-778: which account/team's pins the switcher heads with.
struct PinScope: Hashable {
    let accountId: String
    let teamId: String
}

struct BoardSwitcherSheet: View {
    let boardLoader: MultiAccountBoardLoader?
    let currentBoard: CurrentBoardRef?
    let onSelect: (_ accountId: String, _ boardId: String) -> Void
    /// EXP-698 r5 (Android parity): creating a board and creating a team are
    /// both reachable FROM the switcher — the two things a switcher with
    /// nothing worth switching to should offer.
    let onCreateBoard: () -> Void
    let onCreateTeam: () -> Void
    /// EXP-778: the account + team whose personal pins head the sheet (nil =
    /// no team in view, no section).
    var pinScope: PinScope? = nil
    /// A pinned row was tapped — the caller parks the route and pushes it
    /// once the sheet is gone (the create-entries hand-off).
    var onOpenPinned: (PinnedItem) -> Void = { _ in }

    @Environment(AppDependencies.self) private var deps
    @State private var pinned = PinnedItemsModel()

    var body: some View {
        GlassSheetChrome(title: "Switch board") {
            let groups = boardLoader?.groups ?? []
            LazyVStack(alignment: .leading, spacing: 18) {
                // EXP-778: the sidebar's Pinned section — on top, hidden
                // when the team has no resolvable pin.
                if !pinned.items.isEmpty {
                    pinnedSection
                }
                if groups.isEmpty {
                    emptyHint
                } else {
                    ForEach(groups) { group in
                        serverSection(group)
                    }
                    // At the very bottom, under every server's teams.
                    plainActionRow(
                        icon: AppIcons.uiAdd,
                        title: "New team",
                        identifier: "board-switcher-new-team",
                        action: onCreateTeam
                    )
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 4)
            .padding(.bottom, 24)
        }
        .task(id: pinScope) {
            guard let pinScope else {
                pinned.stop()
                return
            }
            pinned.observe(accountId: pinScope.accountId, teamId: pinScope.teamId, db: deps.db)
        }
    }

    // MARK: - Pinned (EXP-778)

    @ViewBuilder
    private var pinnedSection: some View {
        VStack(alignment: .leading, spacing: 6) {
            HStack(spacing: 8) {
                AppIcon(AppIcons.uiPin, size: 14)
                Text("Pinned")
                    .font(.caption.weight(.semibold))
                Spacer()
            }
            .foregroundStyle(.white.opacity(0.85))
            .padding(.horizontal, 4)

            VStack(spacing: 6) {
                ForEach(pinned.items) { item in
                    Button {
                        onOpenPinned(item)
                    } label: {
                        pinnedRow(item)
                    }
                    .buttonStyle(.plain)
                    .accessibilityIdentifier("pinned-row")
                }
            }
        }
    }

    @ViewBuilder
    private func pinnedRow(_ item: PinnedItem) -> some View {
        HStack(spacing: 12) {
            switch item {
            case let .issue(_, issue):
                AppIcon(AppIcons.uiIssue, size: 16)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                HStack(spacing: 6) {
                    if let identifier = issue.identifier, !identifier.isEmpty {
                        Text(identifier)
                            .font(.caption.monospaced())
                            .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                            .lineLimit(1)
                    }
                    Text(issue.title.isEmpty ? "Untitled issue" : issue.title)
                        .font(.subheadline.weight(.medium))
                        .foregroundStyle(.white)
                        .lineLimit(1)
                        .truncationMode(.tail)
                }
            case let .session(_, session, issue):
                // The sessions list's first line, so the two cannot drift;
                // an ended run wears the static neutral dot.
                let ended = session.status == DomainContract.codingSessionStatusEnded
                SessionRowTitle(
                    identifier: issue?.identifier,
                    title: sessionRowTitle(issue: issue, session: session),
                    state: CodingSessionDisplayState.of(
                        session: session, prState: issue?.prState ?? session.prState
                    ),
                    paused: ended,
                    // EXP-848: pulses only while a turn is open.
                    busy: session.agentBusy
                )
            case let .action(_, action):
                AppIcon(action.icon ?? AppIcons.actionDefault, size: 16)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                Text(action.name)
                    .font(.subheadline.weight(.medium))
                    .foregroundStyle(.white)
                    .lineLimit(1)
            }

            Spacer(minLength: 0)

            AppIcon(AppIcons.uiChevronRight, size: 16)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
        .glassRow()
    }

    private var emptyHint: some View {
        VStack(spacing: 10) {
            AppIcon(AppIcons.navBoards, size: AppIcon.Size.large)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
            Text("Create your first board on the web or desktop app.")
                .font(.caption)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(.horizontal, 40)
        .padding(.vertical, 32)
    }

    @ViewBuilder
    private func serverSection(_ group: ServerBoardGroup) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            // The hostname/email header only disambiguates when several
            // accounts are signed in — with a single account it's noise.
            if (boardLoader?.groups.count ?? 0) > 1 {
                HStack(alignment: .top) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(group.hostname)
                            .font(.subheadline.weight(.semibold))
                            .foregroundStyle(.white.opacity(0.9))
                        if let email = group.userEmail, !email.isEmpty {
                            Text(email)
                                .font(.caption2)
                                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                        }
                    }
                    Spacer()
                }
                .padding(.horizontal, 4)
            }

            ForEach(group.teamBlocks) { block in
                teamBlock(accountId: group.accountId, block: block)
            }
        }
    }

    @ViewBuilder
    private func teamBlock(accountId: String, block: TeamBlock) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            // Just the mark and the name (EXP-698 r5): the board COUNT was a
            // number nobody switches on, and it is right there as the row
            // count underneath.
            HStack(spacing: 8) {
                TeamAvatar(team: block.team, size: 18)
                Text(block.team.name)
                    .font(.caption.weight(.semibold))
                    .foregroundStyle(.white.opacity(0.85))
                Spacer()
            }
            .padding(.horizontal, 4)

            VStack(spacing: 6) {
                ForEach(block.boards) { board in
                    Button {
                        onSelect(accountId, board.id)
                    } label: {
                        boardRow(board, isCurrent: isCurrent(accountId: accountId, boardId: board.id))
                    }
                    .buttonStyle(.plain)
                }

                plainActionRow(
                    icon: AppIcons.uiAdd,
                    title: "Create board",
                    identifier: "board-switcher-create-board",
                    action: onCreateBoard
                )
            }
        }
    }

    private func isCurrent(accountId: String, boardId: String) -> Bool {
        currentBoard?.accountId == accountId && currentBoard?.boardId == boardId
    }

    @ViewBuilder
    private func boardRow(_ board: BoardEntity, isCurrent: Bool) -> some View {
        HStack(spacing: 12) {
            // Board glyph (stored icon, else a shape-derived fallback) tinted
            // with the board color (replaces the plain color dot).
            AppIcon(BoardTypeDisplay.iconName(for: board), size: 16)
                .foregroundStyle(Color(hex: board.color ?? "#888888") ?? .gray)

            Text(board.name)
                .font(.body)
                .foregroundStyle(.white)

            Spacer()

            Text(board.prefix)
                .font(.caption.monospaced())
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))

            // EXP-698 r5: a chevron, not a tick. The ACTIVE row already says
            // so with its brighter fill (`glassRow(isActive:)`), and every row
            // here goes somewhere — which is what a chevron means everywhere
            // else in the app.
            AppIcon(AppIcons.uiChevronRight, size: 16)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 14)
        .glassRow(isActive: isCurrent)
    }

    /// A plain (non-carded) muted row — the two creation entries. Deliberately
    /// card-less: they are not boards, and a glass row here would read as one.
    /// The identifier is what the styleguide lane taps: both titles are also
    /// button LABELS elsewhere on the screen behind the sheet, so a label query
    /// would be ambiguous.
    @ViewBuilder
    private func plainActionRow(
        icon: String,
        title: String,
        identifier: String,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            HStack(spacing: 12) {
                AppIcon(icon, size: 16)
                Text(title)
                    .font(.body)
                Spacer()
            }
            .foregroundStyle(.white.opacity(TextOpacity.secondary))
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(identifier)
    }
}
